src/generators/gradient_based/probe.jl

```diff
@@ -0,0 +1,80 @@
+import ..Objectives: losses_catalogue, distance_l1
+import LinearAlgebra: transpose, Matrix
+import Flux.Losses
+import Distributions: cdf, Normal
+"""
+    ProbeGenerator(; λ::AbstractFloat=0.1, loss::Symbol=:mse, penalty=distance_l1, kwargs...)
+
+Create a generator that generates counterfactual probes using the specified loss function and penalty function.
+
+# Arguments
+- `λ::AbstractFloat`: The regularization parameter for the generator.
+- `loss::Symbol`: The loss function to use for the generator. Defaults to `:mse`.
+- `penalty`: The penalty function to use for the generator. Defaults to `distance_l1`.
+- `kwargs`: Additional keyword arguments to pass to the `Generator` constructor.
+
+# Returns
+A `Generator` object that can be used to generate counterfactual probes.
+
+based on https://arxiv.org/abs/2203.06768
+"""
+function ProbeGenerator(;
+    λ::AbstractFloat=0.1,
+    loss::Symbol=:logitbinarycrossentropy,
+    penalty=distance_l1,
+    kwargs...,
+)
+    @assert haskey(losses_catalogue, loss) "Loss function not found in catalogue."
+    user_loss = losses_catalogue[loss]
+    return Generator(; loss=user_loss, penalty=penalty, λ=λ, kwargs...)
+end
+
+"""
+    invalidation_rate(ce::AbstractCounterfactualExplanation; σ²=0.01, kwargs...)
+
+Calculate the invalidation rate of a counterfactual explanation.
+
+# Arguments
+- `ce::AbstractCounterfactualExplanation`: The counterfactual explanation to calculate the invalidation rate for.
+- `kwargs`: Additional keyword arguments to pass to the function.
+
+# Returns
+The invalidation rate of the counterfactual explanation.
+"""
+function invalidation_rate(ce::AbstractCounterfactualExplanation; kwargs...)
```

Comment:
Since the kwargs don't seem to be used anywhere inside the function body, is it necessary to have them in the parameter list?

Reply:
I've removed it